base/iterator.jl

```diff
@@ -298,10 +298,32 @@ done(it::Repeated, state) = false
 
 repeated(x, n::Int) = take(repeated(x), n)
 
-# product
+
+# Product -- cartesian product of iterators
 
 abstract AbstractProdIterator
+length(p::AbstractProdIterator) = prod(size(p))
+
+# one iterator
+immutable Prod1{I} <: AbstractProdIterator
+    a::I
+end
+product(a) = Prod1(a)
+
+eltype{I}(::Type{Prod1{I}}) = Tuple{eltype(I)}
+size(p::Prod1) = size(p.a)
+
+@inline start(p::Prod1) = (start(p.a),)
```

Comment:
> It doesn't seem necessary to wrap the state in a tuple. I see Zip1 does that too, which I didn't notice before. Should be changed, unless I'm missing something.

Please don't do this again. :-) https://github.com/JuliaLang/julia/issues/13979

But of course here it's not needed.